pkg/controller/core/traits/manualscalertrait/manualscalertrait_controller.go

```diff
@@ -21,48 +21,53 @@ import (
 	"strings"
 
 	cpv1alpha1 "github.com/crossplane/crossplane-runtime/apis/core/v1alpha1"
+	"github.com/crossplane/crossplane-runtime/pkg/event"
 	"github.com/crossplane/crossplane-runtime/pkg/logging"
 	cpmeta "github.com/crossplane/crossplane-runtime/pkg/meta"
 	oamv1alpha2 "github.com/crossplane/oam-kubernetes-runtime/apis/core/v1alpha2"
 	"github.com/crossplane/oam-kubernetes-runtime/pkg/oam"
 	"github.com/go-logr/logr"
 	"github.com/pkg/errors"
-	appsv1 "k8s.io/api/apps/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/discovery"
+	"k8s.io/kube-openapi/pkg/util/proto"
+	"k8s.io/kubectl/pkg/explain"
+	"k8s.io/kubectl/pkg/util/openapi"
 	ctrl "sigs.k8s.io/controller-runtime"
-	"sigs.k8s.io/controller-runtime/pkg/builder"
 	"sigs.k8s.io/controller-runtime/pkg/client"
-	"sigs.k8s.io/controller-runtime/pkg/handler"
-	"sigs.k8s.io/controller-runtime/pkg/predicate"
-	"sigs.k8s.io/controller-runtime/pkg/source"
 
 	"github.com/crossplane/oam-controllers/pkg/oam/util"
 )
 
 // Reconcile error strings.
 const (
-	errLocateWorkload   = "cannot find workload"
-	errLocateResources  = "cannot find resources"
-	errLocateDeployment = "cannot find deployment"
-	errScaleDeployment  = "cannot scale the deployment"
+	errLocateWorkload      = "cannot find workload"
+	errFetchChildResources = "failed to fetch workload child resources"
+	errQueryOpenAPI        = "failed to query openAPI"
+	errScaleResource       = "cannot scale the resource"
 )
 
 // Setup adds a controller that reconciles ContainerizedWorkload.
 func Setup(mgr ctrl.Manager, log logging.Logger) error {
 	reconciler := Reconciler{
-		Client: mgr.GetClient(),
-		Log:    ctrl.Log.WithName("ManualScalarTrait"),
-		Scheme: mgr.GetScheme(),
+		Client:          mgr.GetClient(),
+		DiscoveryClient: *discovery.NewDiscoveryClientForConfigOrDie(mgr.GetConfig()),
+		log:             ctrl.Log.WithName("ManualScalarTrait"),
+		record:          event.NewAPIRecorder(mgr.GetEventRecorderFor("ManualScalarTrait")),
+		Scheme:          mgr.GetScheme(),
 	}
 	return reconciler.SetupWithManager(mgr)
 }
 
 // Reconciler reconciles a ManualScalarTrait object
 type Reconciler struct {
 	client.Client
-	Log    logr.Logger
+	discovery.DiscoveryClient
+	log    logr.Logger
```

Comment:
There's no need to change the log to be private

Reply:
Conventional OO practice recommends not exposing the internal fields. I wonder if there is any reason to make the log public?